Mathlib/Tactic/MkIffOfInductiveProp.lean

```diff
@@ -306,15 +307,23 @@ def mkIffOfInductivePropImpl (ind : Name) (rel : Name) (relStx : Syntax) : MetaM
 
   let (thmTy, shape) ← Meta.forallTelescope type fun fvars ty ↦ do
     if !ty.isProp then throwError "mk_iff only applies to prop-valued declarations"
-    let lhs := mkAppN (mkConst ind univs) fvars
     let fvars' := fvars.toList
     let shape_rhss ← constrs.mapM (constrToProp univs (fvars'.take params) (fvars'.drop params))
     let (shape, rhss) := shape_rhss.unzip
-    pure (← mkForallFVars fvars (mkApp2 (mkConst `Iff) lhs (mkOrList rhss)), shape)
+    if isEq then
+      pure (← mkEq (mkConst ind univs) (← mkLambdaFVars fvars (mkOrList rhss)), shape)
+    else
+      let lhs := mkAppN (mkConst ind univs) fvars
+      pure (← mkForallFVars fvars (mkApp2 (mkConst `Iff) lhs (mkOrList rhss)), shape)
 
   let mvar ← mkFreshExprMVar (some thmTy)
   let mvarId := mvar.mvarId!
-  let (fvars, mvarId') ← mvarId.intros
+  let (fvars, mvarId') ← do
+    if isEq then
+      let (fvars, mvarId') ← mvarId.funextN (params + inductVal.numIndices)
```

Comment:
Note to self: shouldn't we use the existing variable names when doing `funext` here? Not really, the funext is just an internal proof implementation.